File: 3. Behavioral Design Patterns/Iterator/Typescript/IteratorTemplate.ts
// Iterator Pattern – TypeScript Template

interface Iterator<T> { hasNext(): boolean; next(): T }
interface Aggregate<T> { createIterator(): Iterator<T> }

class ConcreteIterator<T> implements Iterator<T> {
  private index = 0
  constructor(private items: T[]) {}
  hasNext(): boolean { return this.index < this.items.length }
  next(): T { return this.items[this.index++] }
}

class ConcreteAggregate<T> implements Aggregate<T> {
  private items: T[] = []
  add(item: T): void { this.items.push(item) }
  createIterator(): Iterator<T> { return new ConcreteIterator(this.items) }
}

;(function main() {
  const aggregate = new ConcreteAggregate<string>()
  aggregate.add('Item1')
  aggregate.add('Item2')
  const iterator = aggregate.createIterator()
  while (iterator.hasNext()) { const item = iterator.next() }
})()

export {}

File: 3. Behavioral Design Patterns/Iterator/C#/IteratorTemplate.cs
// Iterator Pattern – C# Template

using System.Collections.Generic;

namespace IteratorTemplate
{
    interface IIterator<T>
    {
        bool HasNext();
        T Next();
    }

    interface IAggregate<T>
    {
        IIterator<T> CreateIterator();
    }

    class ConcreteIterator<T> : IIterator<T>
    {
        private readonly List<T> items;
        private int index = 0;
        public ConcreteIterator(List<T> items) { this.items = items; }
        public bool HasNext() => index < items.Count;
        public T Next() => items[index++];
    }

    class ConcreteAggregate<T> : IAggregate<T>
    {
        private readonly List<T> items = new();
        public void Add(T item) { items.Add(item); }
        public IIterator<T> CreateIterator() => new ConcreteIterator<T>(items);
    }

    class Program
    {
        static void Main()
        {
            var aggregate = new ConcreteAggregate<string>();
            aggregate.Add("Item1");
            aggregate.Add("Item2");
            var iterator = aggregate.CreateIterator();
            while (iterator.HasNext()) { var item = iterator.Next(); }
        }
    }
}
